source/common/config/watch_map.h

```diff
@@ -0,0 +1,118 @@
+namespace Config {
+
+struct AddedRemoved {
+  AddedRemoved(std::set<std::string>&& added, std::set<std::string>&& removed)
+      : added_(std::move(added)), removed_(std::move(removed)) {}
+  std::set<std::string> added_;
+  std::set<std::string> removed_;
+};
+
+struct Watch {
+  Watch(SubscriptionCallbacks& callbacks) : callbacks_(callbacks) {}
+  SubscriptionCallbacks& callbacks_;
+  std::set<std::string> resource_names_; // must be sorted set, for set_difference.
+};
+
+// NOTE: Users are responsible for eventually calling removeWatch() on the Watch* returned
+//       by addWatch(). We don't expect there to be new users of this class beyond
+//       NewGrpcMuxImpl and DeltaSubscriptionImpl (TODO(fredlas) to be renamed).
+//
+// Manages "watches" of xDS resources. Several xDS callers might ask for a subscription to the same
+// resource name "X". The xDS machinery must return to each their very own subscription to X.
+// The xDS machinery's "watch" concept accomplishes that, while avoiding parallel redundant xDS
+// requests for X. Each of those subscriptions is viewed as a "watch" on X, while behind the scenes
+// there is just a single real subscription to that resource name.
+//
+// This class maintains the watches<-->subscription mapping: it
+// 1) delivers updates to all interested watches, and
+// 2) tracks which resource names should be {added to,removed from} the subscription when the
+//    {first,last} watch on a resource name is {added,removed}.
+//
+// #1 is accomplished by WatchMap's implementation of the SubscriptionCallbacks interface.
+// This interface allows the xDS client to just throw each xDS update message it receives directly
+// into WatchMap::onConfigUpdate, rather than having to track the various watches' callbacks.
+//
+// The information for #2 is returned by updateWatchInterest(); the caller should use it to
+// update the subscription accordingly.
+//
+// A WatchMap is assumed to be dedicated to a single type_url type of resource (EDS, CDS, etc).
+class WatchMap : public SubscriptionCallbacks, public Logger::Loggable<Logger::Id::config> {
+public:
+  WatchMap() = default;
+
+  // Adds 'callbacks' to the WatchMap, with every possible resource being watched.
+  // (Use updateWatchInterest() to narrow it down to some specific names).
+  // Returns the newly added watch, to be used with updateWatchInterest and removeWatch.
+  Watch* addWatch(SubscriptionCallbacks& callbacks);
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in added_.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
+  //    Y will be in removed_.
+  AddedRemoved updateWatchInterest(Watch* watch,
+                                   const std::set<std::string>& update_to_these_names);
```

Comment:
Q: would it have been plausible to do a `addWatchInterest` and `removeWatchInterest` to simplify some of the logic in here (as well as copies)?

Reply:
All of Envoy's config stuff generally does interest as "my interest has changed to this set" rather than deltas. It would mean a rewrite going all the way out to the users of the config machinery. That rewrite could potentially be a good idea, if delta is supposed to be the new way of things, but now wouldn't be the time for that.